Темы: 
- метрики при ранжировании
<!-- - MLSysDes в задачах/хакатонах (задача на ТЗ)  -->

# 1 Задачи ранжирования
Разбивается на 2 подзадачи:
- выбрать/отобрать нужных кандидатов на ранжирование,
- осуществить процесс ранжирования


### 1.1 Источники

[Материалы][1] и [лекция][2] Воронцова

[Yandex ml book статья][3]

[Итмо][4]

<!-- [Ким Фальк???] -->

[1]: http://www.machinelearning.ru/wiki/index.php?title=%D0%9C%D0%B0%D1%88%D0%B8%D0%BD%D0%BD%D0%BE%D0%B5_%D0%BE%D0%B1%D1%83%D1%87%D0%B5%D0%BD%D0%B8%D0%B5_(%D0%BA%D1%83%D1%80%D1%81_%D0%BB%D0%B5%D0%BA%D1%86%D0%B8%D0%B9%2C_%D0%9A.%D0%92.%D0%92%D0%BE%D1%80%D0%BE%D0%BD%D1%86%D0%BE%D0%B2)#.D0.9E.D0.B1.D1.83.D1.87.D0.B5.D0.BD.D0.B8.D0.B5_.D1.80.D0.B0.D0.BD.D0.B6.D0.B8.D1.80.D0.BE.D0.B2.D0.B0.D0.BD.D0.B8.D1.8E

[2]: https://www.youtube.com/watch?v=kQ5PeshAO1w

[3]: https://education.yandex.ru/handbook/ml/article/zadacha-ranzhirovaniya

[4]: https://neerc.ifmo.ru/wiki/index.php?title=%D0%A0%D0%B0%D0%BD%D0%B6%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5

### 1.2 Постановка задачи

X - множество объектов.

$X^{l}  = \{ x_1, ..., x_{l} \}$ - обучающая выборка.

Есть набор пар $(x_{i}, x_{j})$ для которых известно, что  
$x_{i}$ в каком-то смысле лучше/релевантнее $x_{j}$, но это известно не для всех двух элементов из X. (Частичный порядок) 

**Задача**

Построить ранжирующую функцию $a : X \rightarrow \mathbb{R}$ такую, что для тех пар на которых определен порядок $i \prec j  \Rightarrow a(x_i)<a(x_j)$ 




### 1.3 Примеры

#### Задача ранжирования поисковой выдачи
$D$ — коллекция текстовых документов.

$Q$ — множество запросов.

$D_q \subseteq  D$  — множество документов, найденных по запросу q.

$X = Q \times D$  — объектами являются пары (запрос, документ): $x≡(q,d)$,$q∈Q$,$d∈D_q$.

$Y$ — упорядоченное множество рейтингов.

$y:X→Y$  — оценки релевантности, поставленные асессорами (экспертами): чем выше оценка $y(q,d)$ , тем релевантнее документ $d$ по запросу $q$.

Правильный порядок определен только между документами, найденными по одному и тому же запросу $q$: $(q,d)≺(q,\hat{d})⇔y(q,d)<y(q,\hat{d})$
.

Релевантные ответы запросу $q$ — это список документов $d$, упорядоченных с помощью функции ранжирования $a(q,d)$.

# 2 Виды проведения обучения 
- поточечный (pointwise)
- попарный (pairwise)
- списочный (listwise)

Выбор метода зависит от качества ранжирования данных. Теоретически, списочный подход считается наилучшим, однако, на практике, лучше всего работает попарный подход. (источник?)

### 2.1 Поточечный подход

Оценка релевантности $a(q,d)$ считается для каждого объекта независимо от других пар $(q,d)$

Если речь идёт о задаче ранжирования поисковой выдачи, то пусть асессор поставил какую-то оценку $y$ каждой паре (запрос, документ).
$$ (q, d) \rightarrow y$$

Эта оценка и будет предсказываться. 

При этом никак не учитывается, что **нужно предсказать порядок объектов**, а не оценки. Этот подход является простым в том смысле, что в нём используются уже известные методы. Например, можно предсказывать оценки с использованием линейной регрессии и квадратичной ошибки:

$$ \sum_{i=1}^{l}(a(q_i,d_i)−y(q_i,d_i))^2 \rightarrow min$$




Известно, как решать такую задачу, и таким образом будет получена релевантность. Далее по выходам модели можно ранжировать объекты.

### 2.2 Попарный подход

В попарном подходе используются знания об устройстве целевой переменной. Модель строится сведением к минимуму количества дефектных пар, то есть таких, в которых моделью был предсказан неправильный порядок:

$$ \sum_{x_i < x_j}[a(x_j)- a(x_i)<0] \rightarrow min$$



К сожалению, этот функционал дискретный (в него входят индикаторы), поэтому невозможно его минимизировать. Однако можно действовать так же, как и с классификаторами: оценить функционал сверху.

Можно считать, что разница между объектами $a(x_j)−a(x_i)$  — это отступ $M$, и задать некоторую гладкую функцию $L(M)$:

$$ \sum_{x_i < x_j}L(a(x_j)- a(x_i)) \rightarrow min$$



Если использовать функцию как в логистической регрессии 
$$L(M)=log(1+e^{−M}),$$
то полученный метод называется *RankNet*. Затем можно решать задачу, например, с помощью стохастического градиентного спуска.

# 3 Метрики качества ранжирования
- precision@k, 
- recall@k,
- AP@k,
- MAP@k,
- nDCG

В самой простой постановке задачи ранжирования целевая переменная принимает два значения, документ либо релевантен запросу, либо нет:
$$ y(q,d) \in \{0,1\}$$
где $y$ – целевая переменная, $q$ – запрос, $d$ – документ

$d_q^{(i)}$ будет обозначать $i$-й по релевантности документ для запроса $q$.

##### Precision@k
Эта метрика измеряет долю релевантных элементов среди первых k элементов выходного списка

$$Precision@k(q) = \frac{количество\ релевантных\ в\ топе}{k}$$

$$Precision@k(q) = \frac{1}{k}\sum_{i=1}^{k}y(q,d_{q}^{(i)})$$

**Эта метрика измеряет, насколько точны выходные списки, но не учитывает
качество ранжирования.**

Если более релевантные элементы будут ранжироваться выше в списке, точность не изменится.
Метрика не идеальна и хочется измерять как точность, так и качество ранжирования результатов.



Пусть мы получили два ранжированных вывода модели для двух запросов, а их релевантность $y(q_j,d_{q}^{(i)})$ при $k=5$ представлена в таблице.

|$q$|$d_{q}^{(1)}$|$d_{q}^{(2)}$|$d_{q}^{(3)}$|$d_{q}^{(4)}$|$d_{q}^{(5)}$|
|:---:|:---:|:---:|:---:|:---:|:---:|
|$q_1$|1|1|0|0|0|
|$q_2$|0|1|0|0|1|

Для первого запроса точность модели равна $Precision@k(q_1) = \frac{2}{5}$

Для второго запроса точность модели равна $Precision@k(q_2) = \frac{2}{5}$

<!-- <img src="../09/AP_at_K_example.jpg" alt="drawing" width="500"/> -->


##### Average Precision@k

Метрика берет список из $k$ элементов и вычисляет точность из $k$, усредненную по разным значениям $k$. Высокие значения $AP$ достигаются, когда более релевантные элементы находятся в верхней части списка. Для списка размером $k$ формула $AP$ выглядит так:

$$ AP@k(q) = \frac{\sum_{i=1}^{k}(точность\ на\ i,если\ i-й\ элемент\ релевантен\ для\ пользователя)}{общее\ количество\ релевантных\ элементов}$$

$$ AP@k(q) = \frac{\sum_{i=1}^{k}y(q,d_{q}^{(i)})Precision@i(q)}{\sum_{i=1}^{k}y(q,d_{q}^{(i)})}$$

$AP@k(q)$ - **учитывает качество ранжирования, но вычисляется для конкретного пользователя при конкретном запросе.**

А как учесть разные запросы для разных пользователей?

#### $MAP@k$
Эта метрика сначала вычисляет среднюю точность $AP@k(q_i)$ для каждого запроса ($q_i$), а затем усредняет вычисленные значения по набору запросов.

- Вариант 1 использования $MAP@k$
Для несколько запросов для одного пользователя (тестируем модель на одном пользователе и разных запросах)

- Вариант 2 использования $MAP@k$
запрос один, а выдачи для разных пользователей, тогда усреднение идет по группе пользователей для одного фиксированного запроса. (проверка поисковой системы на одном запросе, но разных пользователях) 

- Вариант 3 использования $MAP@k$
Для несколько запросов для разных пользователей собираем ранжированные ответы, тогда усреднение идет по группе пользователей для одного фиксированного запроса.
(цельная проверка поисковой системы на множестве запросов и разных пользователях)

$$MAP@k = \frac{1}{|Q|} \sum_{q \in Q}AP@k(q)$$

Поскольку показатели точности усредняются, метрика $MAP@k$ **учитывает общее
качество ранжирования списка**. Тем не менее она хороша для бинарной релевантности. То есть она эффективно работает только в том случае, если каждый элемент либо релевантен, либо нет. 

Не работает, если релевантность может быть разной. (для разных типов поиска: поиск организации, поиск ответа на вопрос, поиск человека)

**В случае если документы на запросы могут иметь разную релевантность (непрерывная оценка релевантности), то лучше использовать $nDCG$.**

Пусть мы получили четыре ранжированных вывода модели для четырех запросов а их релевантность $y(q_j,d_{q}^{(i)})$ при $k=5$ представлена в таблице.

|$q$|$d_{q}^{(1)}$|$d_{q}^{(2)}$|$d_{q}^{(3)}$|$d_{q}^{(4)}$|$d_{q}^{(5)}$|
|:---:|:---:|:---:|:---:|:---:|:---:|
|$q_1$|0|0|1|0|0|
|$q_2$|1|0|0|1|0|
|$q_3$|0|0|1|1|0|
|$q_4$|0|0|0|0|0|

Для первого списка ответов:

$$\sum_{i=1}^{k}y(q,d_{q}^{(i)}) = 0 + 0 + 1 + 0 + 0 = 1$$

$Precision@1(q) = 0$, $Precision@2(q) = 0$, 
$Precision@3(q) = \frac{1}{3}$,
$Precision@4(q) = \frac{1}{4}$, 
$Precision@5(q) = \frac{1}{5}$

$$\sum_{i=1}^{k}y(q,d_{q}^{(i)})Precision@i(q) = 
    0 \cdot 0 + 
    0 \cdot 0 + 
    \frac{1}{3} \cdot 1 + 
    \frac{1}{4} \cdot 0 + 
    \frac{1}{5} \cdot 0 = \frac{1}{3}
$$

$$AP@5(q_1) = \frac{\frac{1}{3}}{1} = 0.33$$


Для второго списка ответов:

$$\sum_{i=1}^{k}y(q,d_{q}^{(i)}) = 1 + 0 + 0 + 1 + 0 = 2$$

$$\sum_{i=1}^{k}y(q,d_{q}^{(i)})Precision@i(q) = 
  = 1 + \frac{2}{4} = \frac{3}{2} = 1.5
$$

$$AP@5(q_2) = \frac{\frac{3}{2}}{2} = \frac{3}{4} = 0.75$$

Для третьего списка ответов:

$$\sum_{i=1}^{k}y(q,d_{q}^{(i)}) = 0 + 0 + 1 + 1 + 0 = 2$$

$$\sum_{i=1}^{k}y(q,d_{q}^{(i)})Precision@i(q) = 
  \frac{1}{3} + \frac{2}{4} = \frac{5}{6} 
$$

$$AP@5(q_3) = \frac{\frac{5}{6}}{2} = \frac{5}{12} = 0.42$$

Для четвертого списка ответов:

$$AP@5(q_4) = 0$$

$$MAP@k = \frac{\frac{1}{3} + \frac{3}{4} + \frac{5}{12} + 0}{4} = \frac{18}{12*4} = \frac{3}{8} = 0.375$$


<!-- <img src="../09/MAP_at_k_example.jpg" alt="drawing" width="1000"/> -->




3.2 Discounted Cumulative Gain

DCG (Discounted Cumulative Gain, «дисконтированная накопленная ценность»)

оценивает ценность элементов в списке, взяв за основу показатели
релевантности каждого элемента. Показатели для всех элементов суммируются, причем к каждому показателю применяется понижающий коэффициент, который увеличивается от верхних рангов к нижним. Формула выглядит так:

$$DCG_k(q) = \sum_{i=1}^{k}G_q(d_q^{(i)}) \cdot D(i)$$

Этот подход используется, когда оценки релевантности $y$ могут быть вещественными: $y(q,d) \in \mathbb{R}$. **То есть для каждого документа теперь существует градация между релевантностью и нерелевантностью.**

Здесь $G_q(d)$ — функция пользы, а $D$ — функция дисконтирования от позиции. Для этих функций возможны разные вариации.

Например:
- $G_q(d) = 2^{y(q,d)}-1$, $D(i) = \frac{1}{log_2(i+1)}$

- $G_q(d) = y(q,d)$, $D(i) = \frac{1}{i+1}$

Метрика $DCG_k(q)$ — это сумма дробей. Чем более релевантен документ, тем больше числитель в дроби. Знаменатель зависит от позиции документа, он штрафует за то, где находится документ. Если документ очень релевантен, но занимает низкую позицию, то штраф будет большим, если документ релевантен и находится вверху списка, штраф будет маленьким. Таким образом, метрика $DCG$ учитывает и релевантность, и позицию документа. Она достигает максимума, если все релевантные документы находятся в топе списка, причём отсортированные по значению $y$.

*Иногда при реализации поисковой системы может быть понимание, как падает внимание пользователя с ростом позиции в зависимости от типа запроса. В этом случае функция дисконтирования может стать запросозависимой.*

Однако низкое значение метрики $DCG$ не всегда означает, что ранжирование отработало плохо. Могло быть так, что по запросу просто нет релевантных документов, или же их очень мало. Чтобы избавиться от этой проблемы, значение $DCG$ нормируют на эту метрику при идеальном ранжировании, когда документы отсортированы по истинным значениям релевантности.



#### $nDGC$

Метрика $nDCG$ оценивает качество ранжирования выходного списка: она показывает, насколько хорошо выполнено ранжирование по сравнению с идеальным.
Сначала вычисляется $DCG$, а потом результат делиться на идеальный/лучший способ ранжирования ($iDCG$).

$$nDCG@k(q) = \frac{DCG@k(q)}{iDCG@k(q)}$$

$iDCG@k(q)$ - значение $DCG$ при идеальном ранжировании. После нормировки метрика принимает значения от 0 до 1.

Как и всегда, для получения метрики по набору запросов, считают среднее значение $nDCG$. (усредняют по Q)

##### Пример 1 вычисления $DCG$ и $nDCG$ 

Дано множество документов, где каждый документ оценивается от 3 до 0
, где 3 — очень релевантен, а 0 — не релевантен. Пусть таким множеством будет $S={D_1,D_2,D_3,D_4,D_5,D_6}$, где оценка релевантности по опросу пользователей задается(в том же порядке) множеством $R={3,2,3,0,1,2}$.



|$i$|$G_q(d_q^{(i)})$|$log(i+1)$|$\frac{G_q(d_q^{(i)})}{log(i+1)}$|
|:---:|:---:|:---:|:---:|
|1|3|1|3|
|2|2|1.585|1.262|
|3|3|2|1.5|
|4|0|2.322|0|
|5|1|2.585|0.387|
|6|2|2.807|0.712|

In [29]:
from math import log2

In [30]:
gain = [3,2,3,0,1,2]
DCG = 0
for i in range(1, len(gain)+1):
    val = log2(i+1)
    dcg = gain[i-1]/val
    DCG += dcg
    print("for i = %d, gain = %d, log(i+1) = %.3f, dcg = %.3f"% (i, gain[i-1],val, dcg))
print("DCG = %.3f" % (DCG))

for i = 1, gain = 3, log(i+1) = 1.000, dcg = 3.000
for i = 2, gain = 2, log(i+1) = 1.585, dcg = 1.262
for i = 3, gain = 3, log(i+1) = 2.000, dcg = 1.500
for i = 4, gain = 0, log(i+1) = 2.322, dcg = 0.000
for i = 5, gain = 1, log(i+1) = 2.585, dcg = 0.387
for i = 6, gain = 2, log(i+1) = 2.807, dcg = 0.712
DCG = 6.861


In [31]:
igain = sorted(gain, reverse=True)
iDCG = 0
for i in range(1, len(gain)+1):
    val = log2(i+1)
    dcg = igain[i-1]/val
    iDCG += dcg
    print("for i = %d, gain = %d, log(i+1) = %.3f, dcg = %.3f"% (i, igain[i-1],val, dcg))
print("iDCG = %.3f" % (iDCG))

for i = 1, gain = 3, log(i+1) = 1.000, dcg = 3.000
for i = 2, gain = 3, log(i+1) = 1.585, dcg = 1.893
for i = 3, gain = 2, log(i+1) = 2.000, dcg = 1.000
for i = 4, gain = 2, log(i+1) = 2.322, dcg = 0.861
for i = 5, gain = 1, log(i+1) = 2.585, dcg = 0.387
for i = 6, gain = 0, log(i+1) = 2.807, dcg = 0.000
iDCG = 7.141


In [32]:
DCG/iDCG

0.9608081943360617

Итого $nDCG@6=\frac{DCG@6}{iDCG@6}=\frac{6.861}{7.141}=0.961$
.

##### Пример 2 вычисления $DCG$ и $nDCG$ 

Дано 5 документов, где каждый документ оценивается от 5 до 0
, где 5 — очень релевантен, а 0 — не релевантен. Пусть таким множеством будет $S={D_1,D_2,D_3,D_4,D_5}$, где оценка релевантности по опросу пользователей задается(в том же порядке) множеством $R={0,5,1,4,2}$.


In [33]:
gain = [0,5,1,4,2]

In [34]:
DCG = 0
for i in range(1, len(gain)+1):
    val = log2(i+1)
    dcg = gain[i-1]/val
    DCG += dcg
    print("for i = %d, gain = %d, log(i+1) = %.3f, dcg = %.3f"% (i, gain[i-1],val, dcg))
print("DCG = %.3f" % (DCG))

for i = 1, gain = 0, log(i+1) = 1.000, dcg = 0.000
for i = 2, gain = 5, log(i+1) = 1.585, dcg = 3.155
for i = 3, gain = 1, log(i+1) = 2.000, dcg = 0.500
for i = 4, gain = 4, log(i+1) = 2.322, dcg = 1.723
for i = 5, gain = 2, log(i+1) = 2.585, dcg = 0.774
DCG = 6.151


In [35]:
igain = sorted(gain, reverse=True)
iDCG = 0
for i in range(1, len(gain)+1):
    val = log2(i+1)
    dcg = igain[i-1]/val
    iDCG += dcg
    print("for i = %d, gain = %d, log(i+1) = %.3f, dcg = %.3f"% (i, igain[i-1],val, dcg))
print("iDCG = %.3f" % (iDCG))

for i = 1, gain = 5, log(i+1) = 1.000, dcg = 5.000
for i = 2, gain = 4, log(i+1) = 1.585, dcg = 2.524
for i = 3, gain = 2, log(i+1) = 2.000, dcg = 1.000
for i = 4, gain = 1, log(i+1) = 2.322, dcg = 0.431
for i = 5, gain = 0, log(i+1) = 2.585, dcg = 0.000
iDCG = 8.954


In [36]:
DCG/iDCG

0.6869319726735409

**$nDCG$ хорошо работает в большинстве случаев. Главный недостаток этой
метрики в том, что не всегда можно вычислить эталонные показатели релевантности ($iDCG$). **

# Дополнительно


##### $Recall@k(q)$


#### $MRR(q)$

## что не еще важного есть в этой теме 

Для парного обучения используются следующие подходы (со своими функциями потерь LossFunctions)
- RankingSVM
- RankNet
- LambdaRank
- LambdaMART

Списочный подход к обучению. 

Основной источник - Яндекс ML-book


### Как осуществляется отбор кандидатов +/- 
- случайно,
- потом среди Ложноположительных которые (y = -1, a(x) = 1)

### Другие парные функции потерь

- [PairwiseLogisticLoss][1]
- [PairwiseHingeLoss][2]
- [PairwiseMSELoss][3]

[1]: https://www.tensorflow.org/ranking/api_docs/python/tfr/keras/losses/PairwiseLogisticLoss
[2]: https://www.tensorflow.org/ranking/api_docs/python/tfr/keras/losses/PairwiseHingeLoss
[3]: https://www.tensorflow.org/ranking/api_docs/python/tfr/keras/losses/PairwiseMSELoss

### Какие есть особенности деплоя в задачах ранжирования?